asio/include/xtt/asio/context.inl

```diff
@@ -0,0 +1,340 @@
+                    break;
+                }
+            case return_code::WANT_READ:
+                {
+                    boost::asio::async_read(socket_,
+                                            boost::asio::buffer(io_ptr,
+                                                                bytes_requested),
+                                            boost::asio::bind_executor(strand_,
+                                                                       [this, gpk_lookup_callback, assign_id_callback, success_callback, error_callback]
+                                                                       (auto&& ec, auto&& bytes_transferred)
+                                                                       {
+                                                                           if (ec) {
+                                                                               error_callback(return_code::BAD_IO);
+                                                                               return;
+                                                                           }
+
+                                                                           uint16_t bytes_requested = 0;
+                                                                           unsigned char *io_ptr = NULL;
+
+                                                                           return_code current_rc = handshake_ctx_.handle_io(0,   // no bytes written
+                                                                                                                             bytes_transferred,
+                                                                                                                             bytes_requested,
+                                                                                                                             io_ptr);
+
+                                                                           this->run_state_machine(current_rc,
+                                                                                                   bytes_requested,
+                                                                                                   io_ptr,
+                                                                                                   gpk_lookup_callback,
+                                                                                                   assign_id_callback,
+                                                                                                   success_callback,
+                                                                                                   error_callback);
+                                                                       }));
+                    break;
+                }
+            case return_code::WANT_BUILDSERVERATTEST:
+                {
+                    auto suite_spec = handshake_ctx_.get_suite_spec();
```

Comment:
Why are the `suite_spec` and `cert` checked in each `WANT_X` handler?  Can those just be looked up and checked once, and the result stored as instance members?